firebase-sessions/src/main/kotlin/com/google/firebase/sessions/SharedSessionRepository.kt

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2025 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.firebase.sessions
+
+import android.util.Log
+import androidx.datastore.core.DataStore
+import com.google.firebase.annotations.concurrent.Background
+import com.google.firebase.sessions.ProcessDetailsProvider.getProcessName
+import com.google.firebase.sessions.api.FirebaseSessionsDependencies
+import com.google.firebase.sessions.api.SessionSubscriber
+import com.google.firebase.sessions.settings.SessionsSettings
+import javax.inject.Inject
+import javax.inject.Singleton
+import kotlin.coroutines.CoroutineContext
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+
+/** Repository to persist session data to be shared between all app processes. */
+internal interface SharedSessionRepository {
+  fun appBackground()
+
+  fun appForeground()
+}
+
+@Singleton
+internal class SharedSessionRepositoryImpl
+@Inject
+constructor(
+  private val sessionsSettings: SessionsSettings,
+  private val sessionGenerator: SessionGenerator,
+  private val sessionFirelogPublisher: SessionFirelogPublisher,
+  private val timeProvider: TimeProvider,
+  private val sessionDataStore: DataStore<SessionData>,
+  @Background private val backgroundDispatcher: CoroutineContext,
+) : SharedSessionRepository {
+  /** Local copy of the session data. Can get out of sync, must be double-checked in datastore. */
+  private lateinit var localSessionData: SessionData
+
+  init {
+    CoroutineScope(backgroundDispatcher).launch {
+      sessionDataStore.data.collect { sessionData ->
+        localSessionData = sessionData
```

Comment:
I'm not sure if there's a race condition between `localSessionData` and the app state change methods. Would it make sense to set the value if/after the registered subscribers have been notified - which might also help w/ the cold start detection?

Reply:
Discussed offline